docs/general/cairo_precompiles.md

```diff
@@ -0,0 +1,243 @@
+# Cairo Precompiles
+
+Kakarot zkEVM being a Starknet appchain, it is technically possible to run Cairo
+Contracts on Kakarot. The purpose of this document is to explain the design
+behind the Cairo precompiles, which are the Cairo contracts that are deployed on
+Kakarot to provide additional functionality to the users.
+
+## Requirements
+
+From a developer / user perspective, the precompiles should be easy to interact
+with from EVM contracts. As such, we need to provide a way to call the
+precompiles from EVM contracts, and to return the results to the EVM contracts.
+This involves:
+
+- Converting the EVM calldata to Cairo inputs
+- Converting the Cairo outputs to EVM return data
+
+From a security perspective, the precompiles should _never_ produce Cairo errors
+that would cause the transaction to revert. As we cannot catch contract call
+errors in Cairo, we need to ensure that executing a `call_contract` syscall to a
+precompile will never cause the transaction to revert, meaning that:
+
+- The target Cairo addresses should always correspond to a deployed contract
+- The selector of the Cairo function being called should always be present in
+  the contract called
+- The Cairo contract called should never _panic_
+
+From these principles, we can derive the following design.
+
+## Design
+
+Interacting with the Cairo precompiles will only be done by calling specific EVM
+contracts that have been whitelisted for this purpose. This ensures that the
+Cairo precompiles are only called by contracts that have been reviewed and
+approved regarding the security concerns mentioned above.
+
+As such, the execution flow of an EVM message will be as follows:
+
+```mermaid
+flowchart TD
+    A[Solidity contract] -->|CALL| B
+    B{address type}
+    B -->|EVM Contract| D[Execute contract]
+    B -->|EVM Precompiles| E[Execute EVM precompile]
+    B -->|Cairo Precompile| F{Whitelisted caller code?}
+    F --> |yes| G[Call Cairo contract]
+    F --> |no| H[Call fails]
+```
+
+## Implementation
+
+The solidity part for the Cairo precompiles includes a library that allows
+developers to interact with the Cairo precompiles. This library contains methods
+to call a Cairo contract or class, either via `call` or `staticcall`.
+
+> Note: The behavior of high-level `calls` in solidity prevents calling
+> precompiles directly. As such, the library will use the low-level `call` and
+> `staticcall` opcodes to interact with these Cairo precompiles.
+
+```solidity
```

Comment:
should we update the README here with the complete lib? or at least a link to it